// @feathersjs/generators@5.0.23/lib/hook/index.js
import { dirname } from 'path';
import { fileURLToPath } from 'url';
import { prompt, runGenerators } from '@featherscloud/pinion';
import _ from 'lodash';
import { checkPreconditions, initializeBaseContext } from '../commons.js';
// Set __dirname in es module
const __dirname = dirname(fileURLToPath(import.meta.url));
export const generate = (ctx) => Promise.resolve(ctx)
    .then(initializeBaseContext())
    .then(checkPreconditions())
    .then(prompt(({ type, name }) => [
    {
        type: 'input',
        name: 'name',
        message: 'What is the name of the hook?',
        when: !name
    },
    {
        name: 'type',
        type: 'list',
        when: !type,
        message: 'What kind of hook is it?',
        choices: [
            { value: 'around', name: 'Around' },
            { value: 'regular', name: 'Before, After or Error' }
        ]
    }
]))
    .then((ctx) => {
    const { name } = ctx;
    const kebabName = _.kebabCase(name);
    const camelName = _.camelCase(name);
    return {
        ...ctx,
        kebabName,
        camelName
    };
})
    .then(runGenerators(__dirname, 'templates'));
