// @feathersjs/authentication-client@5.0.23/lib/storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageWrapper = exports.MemoryStorage = void 0;
class MemoryStorage {
    constructor() {
        this.store = {};
    }
    getItem(key) {
        return Promise.resolve(this.store[key]);
    }
    setItem(key, value) {
        return Promise.resolve((this.store[key] = value));
    }
    removeItem(key) {
        const value = this.store[key];
        delete this.store[key];
        return Promise.resolve(value);
    }
}
exports.MemoryStorage = MemoryStorage;
class StorageWrapper {
    constructor(storage) {
        this.storage = storage;
    }
    getItem(key) {
        var _a;
        return Promise.resolve((_a = this.storage) === null || _a === void 0 ? void 0 : _a.getItem(key));
    }
    setItem(key, value) {
        var _a;
        return Promise.resolve((_a = this.storage) === null || _a === void 0 ? void 0 : _a.setItem(key, value));
    }
    removeItem(key) {
        var _a;
        return Promise.resolve((_a = this.storage) === null || _a === void 0 ? void 0 : _a.removeItem(key));
    }
}
exports.StorageWrapper = StorageWrapper;
